src/mpmc_queue.ml

```diff
@@ -0,0 +1,245 @@
+
+  However, with just (2), some dequeuers will struggle to return. If many dequeuers 
+  constatly try to pop an element and fail, they will form a chain.
+
+   tl                 hd
+   |                  |
+  [.]-[A]-[B]-[C]-..-[X]
+
+  For A to rollback, B has to rollback first. For B to rollback C has to rollback first.
+
+  [A] is likely to experience a large latency spike. In such a case, it is easier for [A]
+  to do (3) rather than hope all other active dequeuers will unblock it at some point. 
+  Thus, it's worthwile also trying to do (3) periodically.
+
+  Thus, the current policy does (1) for a bit, then (1), (2) with periodic (3). 
+
+  What about burned slots (4)?
+
+  It's present in the literature. Weakly I'm not a fan. If dequeuers are faster to remove 
+  items than enqueuers supply them, slots burned by dequeuers are going to make enqueuers 
+  do even more work.
+
+  # Resizing 
+
+  The queue does not support resizing, but it can be simulated by wrapping it in a 
+  lockfree list. 
+*)
+
+type 'a t = {
+  array : 'a Option.t Atomic.t Array.t;
+  head : int Atomic.t;
+  tail : int Atomic.t;
+  mask : int;
+}
+
+let create ~size_exponent () : 'a t =
+  let size = 1 lsl size_exponent in
+  let array = Array.init size (fun _ -> Atomic.make None) in
+  let mask = size - 1 in
+  let head = Atomic.make 0 in
+  let tail = Atomic.make 0 in
+  { array; head; tail; mask }
+
+(* [spin_threshold] Number of times on spin on a slot before trying an exit strategy. *)
+let spin_threshold = 30
+
+(* [try_other_exit_every_n] There is two strategies that push/pop can take to fix state (
+   to be able to return without completion). Generally, we want to try to do "rollback" more
+   than "push forward", as the latter adds contention to the side that might already not
+   be keeping up. *)
+let try_other_exit_every_n = 10
+let time_to_try_push_forward n = n mod try_other_exit_every_n == 0
+
+(* [ccas] A slightly nicer CAS. Tries without taking microarch lock first. Use on indices. *)
+let ccas cell seen v =
+  if Atomic.get cell != seen then false else Atomic.compare_and_set cell seen v
+
+let push { array; tail; head; mask; _ } item =
+  let tail_val = Atomic.fetch_and_add tail 1 in
+  let index = tail_val land mask in
+  let cell = Array.get array index in
+
+  (* spin for a bit *)
+  let i = ref 0 in
+  while
+    !i < spin_threshold && not (Atomic.compare_and_set cell None (Some item))
+  do
+    i := !i + 1
+  done;
+
+  (* define clean up function *)
+  let rec take_or_rollback nth_attempt =
+    if Atomic.compare_and_set cell None (Some item) then
+      (* succedded to push *)
+      true
+    else if ccas tail (tail_val + 1) tail_val then (* rolled back tail *)
+      false
+    else if
+      time_to_try_push_forward nth_attempt && ccas head tail_val (tail_val + 1)
+    then (* pushed forward head *)
+      false
+    else (* retry *)
+      take_or_rollback (nth_attempt + 1)
```

Comment:
This function can prevent a thread from progressing if other threads are paused which breaks the definition of obstruction-free (and thus is not lock-free).